himbaechel/uarch/gowin/pack.cc

```diff
@@ -640,6 +654,334 @@ struct GowinPacker
+                                        }
+                                        if (ce_net != this_ce_net) {
+                                            if (ci.attrs.count(id_IOBFF)) {
+                                                log_warning("Conflicting OREG CE nets at %s:'%s' vs '%s'\n",
+                                                            ctx->nameOf(&ci), ctx->nameOf(ce_net),
+                                                            ctx->nameOf(this_ce_net));
+                                            }
+                                        }
+                                        if (lsr_net != this_lsr_net) {
+                                            if (ci.attrs.count(id_IOBFF)) {
+                                                log_warning("Conflicting OREG LSR nets at %s:'%s' vs '%s'\n",
+                                                            ctx->nameOf(&ci), ctx->nameOf(lsr_net),
+                                                            ctx->nameOf(this_lsr_net));
+                                            }
+                                        }
+                                        break;
+                                    }
+                                }
+                            } else {
+                                clk_net = this_clk_net;
+                                ce_net = this_ce_net;
+                                lsr_net = this_lsr_net;
+                                reg_type = ff->type;
+                            }
+                        }
+
+                        // create IOLOGIC cell for flipflop
+                        IdString iologic_name = gwu.create_aux_name(ci.name, 1, "_iobff$");
+                        auto iologic_cell = gwu.create_cell(iologic_name, id_IOLOGICO_EMPTY);
+                        new_cells.push_back(std::move(iologic_cell));
+                        iologic_o = new_cells.back().get();
+
+                        // move ports
+                        for (auto &port : ff->ports) {
+                            IdString port_name = port.first;
+                            ff->movePortTo(port_name, iologic_o, port_name != id_D ? port_name : id_D0);
+                        }
+                        if (ctx->verbose) {
+                            log_info("  place FF %s into OBUF %s, make iologic_o %s\n", ctx->nameOf(ff),
+                                     ctx->nameOf(&ci), ctx->nameOf(iologic_o));
+                        }
+                        iologic_o->setAttr(id_HAS_REG, 1);
+                        iologic_o->setAttr(id_OREG_TYPE, ff->type.str(ctx));
+                        cells_to_remove.push_back(ff->name);
+                    }
+                    break;
+                }
+            }
+
+            // output enable reg in IO
+            if (ci.type == id_IOBUF && (ctx->settings.count(id_IOREG_IN_IOB) || ci.attrs.count(id_IOBFF))) {
+                while (1) {
```

Comment:
As I understand it, the `while` loop here is just to get a way of being able to use `break` - imo `do { } while (false);` would convey the semantics here better.